src/lyk.c

```diff
@@ -0,0 +1,94 @@
+#include "lyk.h"
+#include "lobject.h"
+#include "lopcodes.h"
+#include "ldebug.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include "assert.h"
+
+#ifdef USE_YK
+#include <yk.h>
+#endif
+
+/*
+ * Is the instruction `i` the start of a loop?
+ *
+ * YKFIXME: Numeric and Generic loops can be identified by OP_FORLOOP and OP_TFORLOOP opcodes. 
+ * Other loops like while and repeat-until are harder to identify since they are based on OP_JMP instruction.
+ */
+int is_loop_start(Instruction i)
+{
+  return (GET_OPCODE(i) == OP_FORLOOP || GET_OPCODE(i) == OP_TFORLOOP);
+}
+
+#ifdef USE_YK
+inline YkLocation *yk_lookup_ykloc(CallInfo *ci, Instruction *pc)
+{
+  YkLocation *ykloc = NULL;
+  assert(isLua(ci));
+  Proto *p = ci_func(ci)->p;
+  assert(p->code <= pc && pc <= p->code + p->sizecode);
+  if (is_loop_start(*pc))
+  {
+    ykloc = &p->yklocs[pc - p->code];
+  }
+  return ykloc;
+}
+#endif
+
+inline void yk_set_location(Proto *f, Instruction i, int idx)
+{
+#ifdef USE_YK
+  if (is_loop_start(i))
+  {
+    f->yklocs[idx] = yk_location_new();
+  }
+#endif
+}
+
+inline void yk_init_locaction(Proto *f, Instruction i, int idx, int pc)
+{
+#ifdef USE_YK
+  // YKOPT: Reallocating for every instruction is inefficient.
+  f->yklocs = reallocarray(f->yklocs, pc, sizeof(YkLocation));
+  assert(("Expected yklocs to be defined!", f->yklocs != NULL));
```

Comment:
I've never seen `assert` used like this :P

So with the comma, each expression is checked individually like `&&`  (so a string is always `true`)?

Reply:
Also, where is the initial allocation of `f->yklocs`? I don't see it.

Reply:
I never used assert in C before :no_mouth:, I think its the same as `assert(condition && "message");`
But maybe we should use `lua_assert` here instead as that's the main assertion used in the codebase

Reply:
Ah yeah. If lua has it's own assertions, probably best to use those.

Reply:
I'll change to `lua_assert` but its just al alias:
```
#define lua_assert(c)           assert(c)
```

Reply:
done :point_right: b7fa8d2e00362463ed292910d4a6c5bde239c68b